minetest/minetest
# Add dynamic media without loading from disk

**Ready for review.** Implements & closes #12755.

This PR simply adds the ability to send ephemeral media from a file already in memory as a Lua string. Rather than having to save it to disk, have the engine read it, and then delete it from disk again.

[updated doc for `dynamic_add_media` in `lua_api.txt`](https://github.com/GoodClover/minetest/blob/dynamic_add_media-data/doc/lua_api.txt#L6127)
```lua
minetest.dynamic_add_media({
  filename = "foo.png",
  data = "<raw-png-file>",
  -- data implies (requires) ephemeral=true
}, function() end)
```

As a side-effect, this allows the following, which will read `bar.png` from disk but send it to the client as `foo.png`. You could add an error for this, but why remove a free feature that harms no-one?
```lua
minetest.dynamic_add_media({
  filename = "foo.png",
  filepath = modpath.."/bar.png",
})
```

## To do

- [x] Implement the thing.
- [x] Error checking on table passed to `dynamic_add_media`.

## How to test

1. Download the mod [test.zip](https://github.com/minetest/minetest/files/9552715/test.zip)
2. Check out init.lua, it's not long at all.
3. Load up a world, wait 1 second.
4. Observe an upright_sprite added at your location, both textures working.
5. =^)
